File: native/ios/Sources/OfflineSync/ConnectivityMonitor.swift
import Foundation
import Network

/// Moniteur de connectivité réseau pour iOS
/// Détecte les changements de connexion (WiFi, Cellular, etc.)
public class ConnectivityMonitor {
    
    private let monitor = NWPathMonitor()
    private let queue = DispatchQueue(label: "com.vendor.offlinesync.connectivity")
    private var onConnectivityChanged: ((Bool) -> Void)?
    private var currentPath: NWPath?
    
    // MARK: - Initialization
    
    public init() {
        // Observer le path actuel
        monitor.pathUpdateHandler = { [weak self] path in
            self?.currentPath = path
        }
    }
    
    deinit {
        stopMonitoring()
    }
    
    // MARK: - Connection Status
    
    /// Vérifier si le device est connecté à Internet
    public func isOnline() -> Bool {
        guard let path = currentPath ?? monitor.currentPath as NWPath? else {
            return false
        }
        return path.status == .satisfied
    }
    
    /// Vérifier si la connexion est en WiFi
    public func isWifi() -> Bool {
        guard let path = currentPath ?? monitor.currentPath as NWPath? else {
            return false
        }
        return path.usesInterfaceType(.wifi)
    }
    
    /// Vérifier si la connexion est en données mobiles
    public func isCellular() -> Bool {
        guard let path = currentPath ?? monitor.currentPath as NWPath? else {
            return false
        }
        return path.usesInterfaceType(.cellular)
    }
    
    /// Vérifier si la connexion est en Ethernet (iPad avec adaptateur)
    public func isEthernet() -> Bool {
        guard let path = currentPath ?? monitor.currentPath as NWPath? else {
            return false
        }
        return path.usesInterfaceType(.wiredEthernet)
    }
    
    /// Obtenir le type de connexion
    public func getConnectionType() -> String {
        if !isOnline() {
            return "offline"
        } else if isWifi() {
            return "wifi"
        } else if isCellular() {
            return "cellular"
        } else if isEthernet() {
            return "ethernet"
        } else {
            return "other"
        }
    }
    
    // MARK: - Monitoring
    
    /// Démarrer le monitoring de connectivité
    ///
    /// - Parameter callback: Fonction appelée quand la connectivité change
    ///                       Reçoit true si online, false si offline
    public func startMonitoring(callback: @escaping (Bool) -> Void) {
        self.onConnectivityChanged = callback
        
        monitor.pathUpdateHandler = { [weak self] path in
            guard let self = self else { return }
            
            self.currentPath = path
            let isOnline = path.status == .satisfied
            
            // Appeler le callback sur le main thread
            DispatchQueue.main.async {
                self.onConnectivityChanged?(isOnline)
            }
        }
        
        monitor.start(queue: queue)
    }
    
    /// Arrêter le monitoring
    public func stopMonitoring() {
        monitor.cancel()
        onConnectivityChanged = nil
    }
    
    // MARK: - Connection Details
    
    /// Obtenir des informations détaillées sur la connexion
    public func getConnectionInfo() -> [String: Any] {
        guard let path = currentPath ?? monitor.currentPath as NWPath? else {
            return [
                "is_online": false,
                "connection_type": "offline"
            ]
        }
        
        var info: [String: Any] = [
            "is_online": path.status == .satisfied,
            "connection_type": getConnectionType(),
            "is_wifi": path.usesInterfaceType(.wifi),
            "is_cellular": path.usesInterfaceType(.cellular),
            "is_ethernet": path.usesInterfaceType(.wiredEthernet),
            "is_expensive": path.isExpensive,
            "is_constrained": path.isConstrained
        ]
        
        // Ajouter les interfaces disponibles
        var availableInterfaces: [String] = []
        for interface in path.availableInterfaces {
            switch interface.type {
            case .wifi:
                availableInterfaces.append("wifi")
            case .cellular:
                availableInterfaces.append("cellular")
            case .wiredEthernet:
                availableInterfaces.append("ethernet")
            case .loopback:
                availableInterfaces.append("loopback")
            case .other:
                availableInterfaces.append("other")
            @unknown default:
                availableInterfaces.append("unknown")
            }
        }
        info["available_interfaces"] = availableInterfaces
        
        return info
    }
    
    /// Vérifier si la connexion est coûteuse (données mobiles limitées)
    public func isExpensiveConnection() -> Bool {
        guard let path = currentPath ?? monitor.currentPath as NWPath? else {
            return false
        }
        return path.isExpensive
    }
    
    /// Vérifier si la connexion est limitée (Low Data Mode activé)
    public func isConstrainedConnection() -> Bool {
        guard let path = currentPath ?? monitor.currentPath as NWPath? else {
            return false
        }
        return path.isConstrained
    }
    
    /// Vérifier si le VPN est actif
    public func supportsIPv4() -> Bool {
        guard let path = currentPath ?? monitor.currentPath as NWPath? else {
            return false
        }
        return path.supportsIPv4
    }
    
    /// Vérifier le support IPv6
    public func supportsIPv6() -> Bool {
        guard let path = currentPath ?? monitor.currentPath as NWPath? else {
            return false
        }
        return path.supportsIPv6
    }
    
    /// Vérifier si le support DNS est disponible
    public func supportsDNS() -> Bool {
        guard let path = currentPath ?? monitor.currentPath as NWPath? else {
            return false
        }
        return path.supportsDNS
    }
}

File: native/android/src/main/kotlin/com/techparse/offlinesync/ConnectivityMonitor.kt
package com.techparse.offlinesync

import android.content.Context
import android.net.ConnectivityManager
import android.net.Network
import android.net.NetworkCapabilities
import android.net.NetworkRequest
import android.os.Build

/**
 * Moniteur de connectivité réseau
 * Détecte les changements de connexion (WiFi, Mobile, etc.)
 */
class ConnectivityMonitor(private val context: Context) {
    
    private val connectivityManager = context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
    private var networkCallback: ConnectivityManager.NetworkCallback? = null
    private var onConnectivityChanged: ((Boolean) -> Unit)? = null
    
    /**
     * Vérifier si le device est connecté à Internet
     */
    fun isOnline(): Boolean {
        return try {
            val network = connectivityManager.activeNetwork ?: return false
            val capabilities = connectivityManager.getNetworkCapabilities(network) ?: return false
            
            capabilities.hasCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET) &&
            capabilities.hasCapability(NetworkCapabilities.NET_CAPABILITY_VALIDATED)
        } catch (e: Exception) {
            false
        }
    }
    
    /**
     * Vérifier si la connexion est en WiFi
     */
    fun isWifi(): Boolean {
        return try {
            val network = connectivityManager.activeNetwork ?: return false
            val capabilities = connectivityManager.getNetworkCapabilities(network) ?: return false
            
            capabilities.hasTransport(NetworkCapabilities.TRANSPORT_WIFI)
        } catch (e: Exception) {
            false
        }
    }
    
    /**
     * Vérifier si la connexion est en données mobiles
     */
    fun isCellular(): Boolean {
        return try {
            val network = connectivityManager.activeNetwork ?: return false
            val capabilities = connectivityManager.getNetworkCapabilities(network) ?: return false
            
            capabilities.hasTransport(NetworkCapabilities.TRANSPORT_CELLULAR)
        } catch (e: Exception) {
            false
        }
    }
    
    /**
     * Obtenir le type de connexion
     */
    fun getConnectionType(): String {
        return when {
            !isOnline() -> "offline"
            isWifi() -> "wifi"
            isCellular() -> "cellular"
            else -> "other"
        }
    }
    
    /**
     * Démarrer le monitoring de connectivité
     * 
     * @param callback Fonction appelée quand la connectivité change
     *                 Reçoit true si online, false si offline
     */
    fun startMonitoring(callback: (Boolean) -> Unit) {
        this.onConnectivityChanged = callback
        
        // Créer une requête réseau
        val request = NetworkRequest.Builder()
            .addCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET)
            .addCapability(NetworkCapabilities.NET_CAPABILITY_VALIDATED)
            .build()
        
        // Créer le callback
        networkCallback = object : ConnectivityManager.NetworkCallback() {
            override fun onAvailable(network: Network) {
                super.onAvailable(network)
                callback(true)
            }
            
            override fun onLost(network: Network) {
                super.onLost(network)
                callback(false)
            }
            
            override fun onCapabilitiesChanged(
                network: Network,
                capabilities: NetworkCapabilities
            ) {
                super.onCapabilitiesChanged(network, capabilities)
                
                val isValidated = capabilities.hasCapability(NetworkCapabilities.NET_CAPABILITY_VALIDATED)
                val hasInternet = capabilities.hasCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET)
                
                callback(isValidated && hasInternet)
            }
        }
        
        // Enregistrer le callback
        try {
            connectivityManager.registerNetworkCallback(request, networkCallback!!)
        } catch (e: Exception) {
            // En cas d'erreur, ne rien faire
        }
    }
    
    /**
     * Arrêter le monitoring
     */
    fun stopMonitoring() {
        networkCallback?.let { callback ->
            try {
                connectivityManager.unregisterNetworkCallback(callback)
            } catch (e: Exception) {
                // Ignore si déjà unregistered
            }
        }
        networkCallback = null
        onConnectivityChanged = null
    }
    
    /**
     * Obtenir des informations détaillées sur la connexion
     */
    fun getConnectionInfo(): Map<String, Any> {
        val info = mutableMapOf<String, Any>()
        
        info["is_online"] = isOnline()
        info["connection_type"] = getConnectionType()
        info["is_wifi"] = isWifi()
        info["is_cellular"] = isCellular()
        
        try {
            val network = connectivityManager.activeNetwork
            val capabilities = network?.let { connectivityManager.getNetworkCapabilities(it) }
            
            capabilities?.let { cap ->
                info["has_internet"] = cap.hasCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET)
                info["is_validated"] = cap.hasCapability(NetworkCapabilities.NET_CAPABILITY_VALIDATED)
                info["link_downstream_bandwidth_kbps"] = cap.linkDownstreamBandwidthKbps
                info["link_upstream_bandwidth_kbps"] = cap.linkUpstreamBandwidthKbps
            }
        } catch (e: Exception) {
            // Ignorer les erreurs
        }
        
        return info
    }
    
    /**
     * Vérifier si la connexion est mesurée (données mobiles limitées)
     */
    fun isMeteredConnection(): Boolean {
        return try {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                connectivityManager.isActiveNetworkMetered
            } else {
                // Sur les anciennes versions, considérer cellular comme metered
                isCellular()
            }
        } catch (e: Exception) {
            false
        }
    }
}
